### DISCUSSION TITLE
Chart is redrawn as marker click event

### DISCUSSION THREAD

**Discussion by chaeeunDev** [patrykandpatrick/vico#963] [Questions] [answered]
### Question

When I receive an observable value and click on the chart, the value of the selectedIndex is changed and a persistent marker is displayed in the corresponding location, and the chart is redrawn each time i click.

I want to prevent the chart from being redrawn every time I click. Is there a way to redraw only the marker while using the event I made

```
var selectedIndex by remember { mutableStateOf(-1f) }

marker = marker,
markerVisibilityListener = object : CartesianMarkerVisibilityListener {
    override fun onHidden(marker: CartesianMarker) {
//                    TODO("Not yet implemented")
    }

    override fun onShown(
        marker: CartesianMarker,
        targets: List<CartesianMarker.Target>
    ) {
        val x = targets.first().x
        if (opening.value[x.toInt()] > 0) {
            selectedIndex = targets.first().x.toFloat()
        }
    }
},
persistentMarkers = rememberExtraLambda( selectedIndex) { marker at selectedIndex },


```
### Vico version(s)

2.0.0-alpha.28

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @chaeeunDev. Can you describe what you're looking to solve? Why is the chart redraw an issue?

**Reply by chaeeunDev** [reply]
Hi! @Gowsky! 
I wish only the marker would be redrawn when I press the data point, but the entire chart is currently redrawn.
Is it because I use the state value to float the persistent marker in the pressing part?

**Reply by Gowsky** [reply]
I understand, but what is the actual problem, and what is the end goal? **Why** do you need to prevent the chart from being redrawn?

**Reply by chaeeunDev** [reply]
Sorry for the lack of explanation!
if i just use the basic marker, the marker is redrawn on the touched area not the entire chart,
But when used with a persistent marker, every part of the chart is redrawn. I'm just trying to fix the persistent marker position, but it's a little uncomfortable to see because all the charts keep redrawing!
Is it correct that the redrawing part is different when the position of the basic marker and the persistent marker are changed?
But if it's right to redraw the entire chart when the position of the fixed marker changes, I'll look for it another way!

**Reply by Gowsky** [reply]
It sounds like there might be a misunderstanding as to what "redraw" means. The observations you've shared don't seem to match redrawing. You might be referring to the initial animation, which looks like this:

[...](https://github.com/user-attachments/assets/8ecc841c-ba89-4db5-b189-fe1ba96b3395)

Generally, this shouldn't occur when a persistent marker is moved, but it might happen if there's an issue with your marker logic. Is this animation what you're referring to? If so, please share your code so we can check what's happening.

**Reply by chaeeunDev** [reply]
What you're explaining is what I'm trying to say!
I'm sorry if I misinterpreted it
The animation of the chart is going on again.

The code is as follows
```
var selectedIndex by remember { mutableStateOf(-1f) }

val candle = rememberCandle(
        body = rememberLineComponent(
            color = MaterialTheme.colorScheme.primary,
            shape = Shape.rounded(60),
            thickness = thickness
        )
    )

    val marker = rememberMarker(valueFormatter = BasicCandleStickChartFormatter())

    CartesianChartHost(
        CartesianChart(
            rememberCandlestickCartesianLayer(
                candles = CandlestickCartesianLayer.CandleProvider.absolute(
                    candle,
                    candle,
                    candle
                ),
                candleSpacing = spacing,
                axisValueOverrider = axisValueOverrider
            ),
            endAxis = rememberEndAxis(
                line = null,
                rememberAxisLabelComponent(color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)),
            ),
            bottomAxis =
            rememberBottomAxis(
                line = rememberAxisLineComponent(),
                rememberAxisLabelComponent(color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)),
                guideline = null,
                tick = null,
                valueFormatter =
                { x, chartValues, _ ->
//                    Log.d("codeflow", "formatting value from candlestick with $x and $chartValues")
                    if (x.toInt() < size) {
                        x_labelList[x.toInt()]
//                            chartValues.model.extraStore[labelListKey][x.toInt()]
//                        x.toString()
                    } else {
                        x.toString()
                    }
                },
                itemPlacer = HorizontalAxis.ItemPlacer.default(spacing = 1, addExtremeLabelPadding = true),
            ),
            horizontalLayout = HorizontalLayout.FullWidth(),
            marker = marker,
            // vico github Discussions #963
            markerVisibilityListener = object : CartesianMarkerVisibilityListener {
                override fun onHidden(marker: CartesianMarker) {
//                    TODO("Not yet implemented")
                }

                override fun onShown(
                    marker: CartesianMarker,
                    targets: List<CartesianMarker.Target>
                ) {
                    val x = targets.first().x
                    if (opening.value[x.toInt()] > 0) {
                        selectedIndex = targets.first().x.toFloat()
                    }
                }
            },
            persistentMarkers = rememberExtraLambda( selectedIndex) { marker at selectedIndex },
        ),

//        scrollState = scrollState,
//        zoomState = zoomState,
        modelProducer = modelProducer,
    )
    ```

**Reply by Gowsky** [reply, marked as answer]
The problem is caused by the use of the `CartesianChart` constructor instead of `rememberCartesianChart`. Please update your code as follows to fix it:

```kt
CartesianChartHost(rememberCartesianChart(...), ...)
```

Also:

- From your formatter:

  ```kt
  { x, chartValues, _ ->
      // Log.d("codeflow", "formatting value from candlestick with $x and $chartValues")
      if (x.toInt() < size) {
          x_labelList[x.toInt()]
          // chartValues.model.extraStore[labelListKey][x.toInt()]
          // x.toString()
      } else {
          x.toString()
      }
  }
  ```

  The first line in the `if` block is incorrect. You should use the second one, which is commented out. Extras are required here for proper synchronization. The same applies to `size`, but instead of converting it to an extra, you can remove it and use `getOrNull` for simplicity. You should also use `remember`. The final version looks like this:

  ```kt
  remember {
      { x, chartValues, _ ->
          chartValues.model.extraStore[labelListKey].getOrNull(x.toInt()) ?: x.toString()
      }
  }
  ```

- `remember` is missing in a few more places. You shoud make these changes:

  ```kt
  remember { BasicCandleStickChartFormatter() }
  ```

  ```kt
  remember { HorizontalAxis.ItemPlacer.default(...) }
  ```

  ```kt
  remember {
      object : CartesianMarkerVisibilityListener {
          ...
      }
  }
  ```

**Reply by chaeeunDev** [reply]
Oh, the problem was that I didn't use "remember"
When I applied it to the code right away, the problem was solved!!!!!
 Thank you for fixing not only this but also other codes!!

**Reply by Gowsky** [reply]
You're welcome!
